task/button/src/main.rs

```diff
@@ -0,0 +1,271 @@
+            Ok(self.increment()?)
+        }
+    }
+}
+
+impl idl::InOrderButtonImpl for ButtonServer {
+    /// Simulate a button press
+    fn press(
+        &mut self,
+        _: &userlib::RecvMessage,
+    ) -> Result<u8, RequestError<ButtonError>> {
+        self.handle_button_press()?;
+        Ok(self.rgb)
+    }
+
+    fn off(
+        &mut self,
+        _: &userlib::RecvMessage,
+    ) -> Result<(), RequestError<ButtonError>> {
+        self.rgb = 0;
+        let _ = self.update_leds()?;
+        Ok(())
+    }
+
+    fn set(
+        &mut self,
+        _: &userlib::RecvMessage,
+        rgb: u8,
+    ) -> Result<(), RequestError<ButtonError>> {
+        if rgb >= 8 {
+            Err(ButtonError::InvalidValue.into())
+        } else {
+            self.rgb = rgb % 8;
+            let _ = self.update_leds()?;
+            Ok(())
+        }
+    }
+
+    fn blink(
+        &mut self,
+        _: &userlib::RecvMessage,
+        on: u32,
+        off: u32,
+    ) -> Result<(), RequestError<ButtonError>> {
+        self.on_ms = on;
+        self.off_ms = off;
+        self.on = true;
+        set_timer_relative(self.on_ms, notifications::TIMER_MASK);
+        let _ = self.update_leds()?;
+        Ok(())
+    }
+}
+
+impl NotificationHandler for ButtonServer {
+    fn current_notification_mask(&self) -> u32 {
+        notifications::TIMER_MASK + notifications::BUTTON_IRQ_MASK
+    }
+
+    fn handle_notification(&mut self, bits: u32) {
+        if (bits & notifications::TIMER_MASK) != 0 {
+            self.timer_expiry()
+        }
+
+        if (bits & notifications::BUTTON_IRQ_MASK) != 0 {
+            if let Ok(detected) = self.gpio.detected_falling(BUTTON_PINT_MASK) {
+                self.gpio.clear_falling(BUTTON_PINT_MASK);
+                self.gpio.clear_status(BUTTON_PINT_MASK);
+                if detected != 0 {
+                    let _ = self.handle_button_press();
+                }
+            } else {
+                self.gpio.clear_falling(BUTTON_PINT_MASK);
+                self.gpio.clear_status(BUTTON_PINT_MASK);
+            }
+            sys_irq_control(notifications::BUTTON_IRQ_MASK, true);
+        }
+    }
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    // XXX Safety, if all else fails, then waiting a minute longer after reset
+    // may help us get a debugger in there.
+    // hl::sleep_for(60000);
+    let mut buffer = [0u8; INCOMING_SIZE];
+
+    let gpio_driver = GPIO.get_task_id();
+    setup_pins(gpio_driver).unwrap_lite();
+    let gpio = drv_lpc55_gpio_api::Pins::from(gpio_driver);
+
+    let mut server = ButtonServer {
+        gpio,
+        quick: 0,
+        last_button_press: 0,
+        on: true, // LEDs are on
+        on_ms: 0,
+        off_ms: 0,  // timer inactive, with on_ms > 0, increment
+        rgb: 0b111, // start with all LEDs on
+    };
+
+    // XXX for now at least, setting up a PINT pin will set dir to Input
+    server.gpio.disable_rising(BUTTON_PINT_MASK);
```

Comment:
Some of these calls are redundant with the reset state of the PINT peripheral and can be removed.